File: participants/anwesha-patro/week1/src/controllers/productController.js
// const AppError = require("../utils/appError");
// const { successResponse } = require("../utils/apiResponse");
// const Product = require("../models/product.mongo"); // your Mongoose model

// // GET all products
// exports.getAllProducts = async (req, res, next) => {
//     try {
//         const products = await Product.find();
//         return successResponse(res, "All products fetched successfully", products);
//     } catch (error) {
//         return next(new AppError(error.message || "Failed to fetch products", 500));
//     }
// };

// // GET product by ID
// exports.getProductById = async (req, res, next) => {
//     try {
//         const { id } = req.params;
//         const product = await Product.findById(id);

//         if (!product) {
//             return next(new AppError("Product not found", 404));
//         }

//         return successResponse(res, "Product fetched successfully", product);
//     } catch (error) {
//         return next(new AppError(error.message || "Failed to fetch product", 500));
//     }
// };

// // CREATE product (POST)
// exports.createProduct = async (req, res, next) => {
//     try {
//         const newProduct = await Product.create(req.body);
//         return successResponse(res, "Product created successfully", newProduct, 201);
//     } catch (error) {
//         return next(new AppError(error.message || "Failed to create product", 500));
//     }
// };

// // UPDATE product (PUT)
// exports.updateProduct = async (req, res, next) => {
//     try {
//         const { id } = req.params;
//         const updatedProduct = await Product.findByIdAndUpdate(id, req.body, {
//             new: true,
//             runValidators: true,
//         });

//         if (!updatedProduct) {
//             return next(new AppError("Product not found", 404));
//         }

//         return successResponse(res, "Product updated successfully", updatedProduct);
//     } catch (error) {
//         return next(new AppError(error.message || "Failed to update product", 500));
//     }
// };

// // PATCH product
// exports.patchProduct = async (req, res, next) => {
//     try {
//         const { id } = req.params;
//         const patchedProduct = await Product.findByIdAndUpdate(id, req.body, {
//             new: true,
//             runValidators: true,
//         });

//         if (!patchedProduct) {
//             return next(new AppError("Product not found", 404));
//         }

//         return successResponse(res, "Product updated partially", patchedProduct);
//     } catch (error) {
//         return next(new AppError(error.message || "Failed to patch product", 500));
//     }
// };

// // DELETE product
// exports.deleteProduct = async (req, res, next) => {
//     try {
//         const { id } = req.params;
//         const deletedProduct = await Product.findByIdAndDelete(id);

//         if (!deletedProduct) {
//             return next(new AppError("Product not found", 404));
//         }

//         return successResponse(res, "Product deleted successfully", deletedProduct);
//     } catch (error) {
//         return next(new AppError(error.message || "Failed to delete product", 500));
//     }
// };


// controllers/productController.js
const productService = require("../services/productService");
const AppError = require("../utils/appError");
const { successResponse } = require("../utils/apiResponse");

// -------------------- GET ALL PRODUCTS --------------------
exports.getAllProducts = async (req, res, next) => {
    try {
        const products = await productService.getAllProducts();
        // Ensure response shows both DBs clearly
        const response = {
            mongo: products.mongo || [],
            sql: products.sql || [],
        };
        return successResponse(res, "All products fetched successfully", response);
    } catch (err) {
        return next(new AppError(err.message || "Failed to fetch products", 500));
    }
};

// -------------------- GET PRODUCT BY ID --------------------
exports.getProductById = async (req, res, next) => {
    try {
        const product = await productService.getProductById(req.params.id);
        if (!product || (!product.mongo && !product.sql)) {
            return next(new AppError("Product not found", 404));
        }
        const response = {
            mongo: product.mongo || null,
            sql: product.sql || null,
        };
        return successResponse(res, "Product fetched successfully", response);
    } catch (err) {
        return next(new AppError(err.message || "Failed to fetch product", 500));
    }
};

// -------------------- CREATE PRODUCT (POST) --------------------
exports.createProduct = async (req, res, next) => {
    try {
        const product = await productService.createProduct(req.body);
        const response = {
            mongo: product.mongo || null,
            sql: product.sql || null,
        };
        return successResponse(res, "Product created successfully", response, 201);
    } catch (err) {
        return next(new AppError(err.message || "Failed to create product", 500));
    }
};

// -------------------- UPDATE PRODUCT (PUT) --------------------
exports.updateProduct = async (req, res, next) => {
    try {
        const product = await productService.updateProduct(req.params.id, req.body);
        if (!product || (!product.mongo && !product.sql)) {
            return next(new AppError("Product not found", 404));
        }
        const response = {
            mongo: product.mongo || null,
            sql: product.sql || null,
        };
        return successResponse(res, "Product updated successfully", response);
    } catch (err) {
        return next(new AppError(err.message || "Failed to update product", 500));
    }
};

// -------------------- PATCH PRODUCT --------------------
exports.patchProduct = async (req, res, next) => {
    try {
        const product = await productService.patchProduct(req.params.id, req.body);
        if (!product || (!product.mongo && !product.sql)) {
            return next(new AppError("Product not found", 404));
        }
        const response = {
            mongo: product.mongo || null,
            sql: product.sql || null,
        };
        return successResponse(res, "Product updated partially", response);
    } catch (err) {
        return next(new AppError(err.message || "Failed to patch product", 500));
    }
};

// -------------------- DELETE PRODUCT --------------------
exports.deleteProduct = async (req, res, next) => {
    try {
        const product = await productService.deleteProduct(req.params.id);
        if (!product || (!product.mongo && !product.sql)) {
            return next(new AppError("Product not found", 404));
        }
        const response = {
            mongo: product.mongo || null,
            sql: product.sql || null,
        };
        return successResponse(res, "Product deleted successfully", response);
    } catch (err) {
        return next(new AppError(err.message || "Failed to delete product", 500));
    }
};

File: templates/week1/day1-node-rest-api/participants/abhinendra/FAST_API/ECOMMERCE_FAST_API/app/controllers/product_controller.py
from fastapi import APIRouter, HTTPException, Depends

from app.services.product_service import get_all_products,get_product_by_id,create_product,update_product_by_id,patch_update,delete_product_by_id

from app.models.product_model import Product,patch_product

from sqlalchemy.orm import Session

from app.core.config import get_db

from sqlalchemy.ext.asyncio import AsyncSession

router = APIRouter(
    prefix="/api/products",
    tags=["products"]
)

@router.get("/")
async def get_products(db : AsyncSession = Depends(get_db)):
    return await get_all_products(db)



#GET products by ID
@router.get("/{product_id}")
async def get_product(product_id: int,db : AsyncSession = Depends(get_db)):
    product = await get_product_by_id(db , product_id)

    if not product:
        raise HTTPException(status_code= 404, detail= "Product not Found")

    return product


#CREATE product
@router.post("/")
async def add_product(product : Product,db : AsyncSession = Depends(get_db)):
    return await create_product(db , product)


#update product using put

@router.put("/{product_id}")
async def update_product(product_id: int, product: Product, db : AsyncSession = Depends(get_db)):
    updated_product = await update_product_by_id(db, product_id, product)

    if updated_product is None:
        raise HTTPException(status_code=404, detail="Product not found")

    return updated_product

#update partial product
@router.patch("/{product_id}")
async def update_partial_product(
    product_id:int, 
    patch_data: patch_product,
    db : AsyncSession = Depends(get_db)
):
    patched_product = await patch_update(product_id, patch_data,db)

    if not patched_product:
        raise HTTPException(status_code=404, detail="Product not found")
    
    return patched_product
# @router.patch("/{product_id}")
# def update_partial_product(
#     product_id: int,
#     patch_data: patch_product,
#     db: Session = Depends(get_db)
# ):
#     patched_product = patch_update(product_id, patch_data, db)

#     if not patched_product:
#         raise HTTPException(status_code=404, detail="Product not found")
    
#     return patched_product


#delete product
@router.delete("/{product_id}")
async def delete_product(product_id: int,db : AsyncSession = Depends(get_db)):
    deleted_product = await delete_product_by_id(product_id)

    if deleted_product is None:
        raise HTTPException(status_code=404, detail="Product not found")

    return {
        "message": "Product deleted successfully",
        "product": deleted_product
    }